### PR TITLE
Container instances/groups management client

### PR DESCRIPTION
**Pull request by ScottHolden**
Generated the ContinerInstances client along with a base set of scenario tests.

This rest api spec was already merged, so generated it straight off there :)

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [X] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [X] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [X] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [X] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

**Comment by shahabhijeet** [comment]
@sauryadas @rgardler Can one of you review this PR for Container instance.

**Comment by ScottHolden** [comment]
Hi @shahabhijeet ,
This was merged into the rest spec back in https://github.com/Azure/azure-rest-api-specs/pull/1480

The .NET SDK did not have this client, thus it was created from scratch.

Let me know if you need any more info :)

**Comment by shahabhijeet** [comment]
@sauryadas can you approve this PR?

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/Utilities/TestContext.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+using Microsoft.Azure.Test.HttpRecorder;
+using Microsoft.Rest.ClientRuntime.Azure.TestFramework;
+
+namespace ContainerInstance.Tests
+{
+	class TestContext : IDisposable
+	{
+		public const string DefaultTestPrefix = "defaulttestprefix-";
+		public const string DefaultResourceGroupPrefix = "testresourcegroup-";
+		public const string DefaultLocationId = "japaneast";
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/Utilities/TestContext.cs]: >public const string DefaultLocationId = "japaneast"; [](start = 2, length = 52)

@scott@nullops.io No need. Just use location.

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/Utilities/TestContext.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		public const string DefaultLocation = "Japan East";
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/Utilities/TestContext.cs]: >"Japan East"; [](start = 40, length = 13)

Is it possible to use "westus" instead?

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/Utilities/TestContext.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		{
+			try
+			{
+				return HttpMockServer.GetAssetName(methodName, prefix);
+			}
+			catch (KeyNotFoundException e)
+			{
+				throw new KeyNotFoundException(string.Format("Generated name not found for calling method: {0}", methodName), e);
+			}
+		}
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/Utilities/TestContext.cs]: Can you just use `TestUtilities.GenerateName`

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		public void ContainerGroupCreateGetDeleteTests()
+		{
+			using (TestContext context = new TestContext(this))
+			{
+				string containerInstanceLocation = "westus";
+
+				ResourceGroup resourceGroup = context.CreateResourceGroup("containergroupcrd-", containerInstanceLocation);
+				ContainerInstanceManagementClient containerClient = context.GetClient<ContainerInstanceManagementClient>();
+
+				string containerGroupName = context.GenerateName("containergroup");
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs]: >context.GenerateName [](start = 32, length = 20)

Just use `TestUtilities.GenerateName`

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+				string containerOsType = "Linux";
+
+				string containerName = "test1";
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs]: >"test1" [](start = 27, length = 7)

Can you also use `TestUtilities.GenerateName` to generate a random name?

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		{
+			using (TestContext context = new TestContext(this))
+			{
+				string containerInstanceLocation = "westus";
+
+				ResourceGroup resourceGroup = context.CreateResourceGroup("containergroupcrd-", containerInstanceLocation);
+				ContainerInstanceManagementClient containerClient = context.GetClient<ContainerInstanceManagementClient>();
+
+				string containerGroupName = context.GenerateName("containergroup");
+				string containerOsType = "Linux";
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs]: >"Linux" [](start = 29, length = 7)

Use `OperatingSystemTypes.Linux`

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+									MemoryInGB = containerMemoryInGB
+								}
+							}
+						}
+					}
+				});
+				Assert.NotNull(createdContainerGroup);
+				Assert.Equal(containerGroupName, createdContainerGroup.Name);
+				Assert.Equal(containerOsType, createdContainerGroup.OsType);
+				Assert.Equal(1, createdContainerGroup.Containers.Count);
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs]: Also verify the there is public IP address assigned.

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+				Assert.Equal(containerGroupName, getContainerGroup.Name);
+				Assert.Equal(containerOsType, getContainerGroup.OsType);
+
+				Container getContainer = getContainerGroup.Containers.First();
+				Assert.Equal(containerName, getContainer.Name);
+				Assert.Equal(containerImage, getContainer.Image);
+				Assert.Equal(1, getContainer.Ports.Count);
+				Assert.Equal(containerPort, getContainer.Ports.First().Port);
+				Assert.Equal(containerCpu, getContainer.Resources.Requests.Cpu);
+				Assert.Equal(containerMemoryInGB, getContainer.Resources.Requests.MemoryInGB);
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs]: Wrap all the assertion into a helper method.

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+				IPage<ContainerGroup> listRgContainerGroup = containerClient.ContainerGroups.ListByResourceGroup(resourceGroup.Name);
+				Assert.NotNull(listRgContainerGroup);
+				Assert.Equal(1, listRgContainerGroup.Count());
+				Assert.Equal(containerGroupName, listRgContainerGroup.First().Name);
+
+				// List the container group within the subscription
+				IPage<ContainerGroup> listSubContainerGroup = containerClient.ContainerGroups.List();
+				Assert.NotNull(listRgContainerGroup);
+				Assert.True(listRgContainerGroup.Count() >= 1);
+				Assert.True(listRgContainerGroup.Any(x => containerGroupID.Equals(x?.Id)));
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs]: Create a test case in which you create 2 container group. When listing all, you should get both. When listing by resource group, you should only get one.

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+				Container createdContainer = createdContainerGroup.Containers.First();
+				Assert.Equal(containerName, createdContainer.Name);
+				Assert.Equal(containerImage, createdContainer.Image);
+				Assert.Equal(1, createdContainer.Ports.Count);
+				Assert.Equal(containerPort, createdContainer.Ports.First().Port);
+				Assert.Equal(containerCpu, createdContainer.Resources.Requests.Cpu);
+				Assert.Equal(containerMemoryInGB, createdContainer.Resources.Requests.MemoryInGB);
+
+				// Wait till the container group is ready before proceeding
+				ContainerGroupUtilities.WaitTillProvisioningStateSucceeded(containerClient, resourceGroup.Name, containerGroupName);
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerGroupScenarioTests.cs]: Not really necessary. You can call get or list before reach succeeded provisioning state.

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/Utilities/ContainerInstanceTestUtilities.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace ContainerInstance.Tests.Utilities
+{
+    class ContainerInstanceTestUtilities
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/Utilities/ContainerInstanceTestUtilities.cs]: >ContainerInstanceTestUtilities [](start = 10, length = 30)

Seems not used anywhere? Remove this class?

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerInstance.Tests.csproj [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    <PackageReference Include="Microsoft.Azure.Management.Network" Version="11.1.0-preview" />
+    <PackageReference Include="Microsoft.Azure.Management.RecoveryServices" Version="4.2.0-preview" />
+    <PackageReference Include="Microsoft.Azure.Management.RecoveryServices.Backup" Version="1.2.0-preview" />
+    <PackageReference Include="Microsoft.Azure.Management.ResourceManager" Version="1.6.0-preview" />
+    <PackageReference Include="Microsoft.Azure.Management.Storage" Version="6.5.0-preview" />
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="1.1.2" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="1.1.2" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="1.1.2" />
+    <PackageReference Include="WindowsAzure.Storage" Version="8.1.4" />
+  </ItemGroup>
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerInstance.Tests.csproj]: Remove. All not needed.

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerLogScenarioTests.cs [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+			using (TestContext context = new TestContext(this))
+			{
+				string containerInstanceLocation = "westus";
+
+				ResourceGroup resourceGroup = context.CreateResourceGroup("containerlogget-", containerInstanceLocation);
+				ContainerInstanceManagementClient containerClient = context.GetClient<ContainerInstanceManagementClient>();
+
+				string containerGroupName = context.GenerateName("containergroup");
+				string containerOsType = "Linux";
+				string containerName = "test1";
```
Comment (yolo3301) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerLogScenarioTests.cs]: Similar to the other scenario tests. Use built in function to generate name.

**Review by yolocs — CHANGES_REQUESTED** [review]
Hi Scott, thanks for contributing. Left several comments.

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerInstance.Tests.csproj [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  <PropertyGroup>
+    <PackageId>ContainerInstance.Tests</PackageId>
+    <AssemblyName>ContainerInstance.Tests</AssemblyName>
+    <VersionPrefix>1.0.0-preview</VersionPrefix>
+  </PropertyGroup>
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp1.1</TargetFrameworks>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Remove="Utilities\ContainerInstanceTestUtilities.cs" />
```
Comment (shahabhijeet) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerInstance.Tests.csproj]: @ScottHolden this is not needed. I know VS might have added it, can you remove it. VS will include code files by default in all sub directories.

File: src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerInstance.Tests.csproj [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
+
+  <ItemGroup>
+    <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Folder Include="SessionRecords\" />
```
Comment (shahabhijeet) [inline src/SDKs/ContainerInstance/ContainerInstance.Tests/ContainerInstance.Tests.csproj]: @ScottHolden can you clean up these itemgroup, you do not need a separate directory itemgroup for SessionRecords

**Review by shahabhijeet — CHANGES_REQUESTED** [review]
@ScottHolden thank you for your valuable contribution. This is helpful.
Please make changes suggested in the PR

**Comment by shahabhijeet** [comment]
@ScottHolden this PR has not addressed the feedback and the below PR is ready to be merged
https://github.com/Azure/azure-sdk-for-net/pull/3615
Closing this PR

**Pull request closed without merging**